<h1>Исследование надёжности заёмщиков<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков-" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков--2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков <a name="Шаг 2.1 Заполнение пропусков"></a></a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления<a name="4.2  Шаг 2.2 Проверка данных на аномалии и исправления"></a></a></span><ul class="toc-item"><li><span><a href="#Проверка-и-исправление-столбца-children" data-toc-modified-id="Проверка-и-исправление-столбца-children-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проверка и исправление столбца children</a></span></li><li><span><a href="#Проверка-и-исправление-столбцов-dob_years-и-income_type" data-toc-modified-id="Проверка-и-исправление-столбцов-dob_years-и-income_type-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка и исправление столбцов dob_years и income_type</a></span></li><li><span><a href="#Проверка-и-исправление-столбцов-education-и-education_id" data-toc-modified-id="Проверка-и-исправление-столбцов-education-и-education_id-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Проверка и исправление столбцов education и education_id</a></span></li><li><span><a href="#Проверка-и-исправление-столбцов-family_status-и-family_status_id" data-toc-modified-id="Проверка-и-исправление-столбцов-family_status-и-family_status_id-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Проверка и исправление столбцов family_status и family_status_id</a></span></li><li><span><a href="#Проверка-и-исправление-столбца-gender" data-toc-modified-id="Проверка-и-исправление-столбца-gender-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Проверка и исправление столбца gender</a></span></li><li><span><a href="#Проверка-и-исправление-столбцa-purpose" data-toc-modified-id="Проверка-и-исправление-столбцa-purpose-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Проверка и исправление столбцa purpose</a></span></li><li><span><a href="#Проверка-и-исправление-столбцов-days_employed-и-total_income" data-toc-modified-id="Проверка-и-исправление-столбцов-days_employed-и-total_income-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Проверка и исправление столбцов days_employed и total_income</a></span></li></ul></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:--Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:--Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Вопрос 1:  Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Вывод 1:</a></span></li><li><span><a href="#Вопрос-2:--Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:--Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Вопрос 2:  Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>Вывод 2:</a></span></li><li><span><a href="#Вопрос-3:--Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:--Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-9.5"><span class="toc-item-num">9.5&nbsp;&nbsp;</span>Вопрос 3:  Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-9.6"><span class="toc-item-num">9.6&nbsp;&nbsp;</span>Вывод 3:</a></span></li><li><span><a href="#Вопрос-4:--Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:--Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-9.7"><span class="toc-item-num">9.7&nbsp;&nbsp;</span>Вопрос 4:  Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-9.8"><span class="toc-item-num">9.8&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Шаг 1. Обзор данных

In [1]:
import pandas as pd #импорт библиотеки pandas для работы с датафреймом
import os #Для загрузки данных с разных источников

Объявим переменную **static_borrower**, запишем туда датафрейм **data.csv**, который хранит статистику о платёжеспособности клиентов и выведем его на экран.

In [2]:
#Загрузка файла с данными
pth1 = '/datasets/data.csv'
pth2 = 'data.csv'
if os.path.exists(pth1):
    static_borrower = pd.read_csv(pth1)
elif os.path.exists(pth2):
    static_borrower = pd.read_csv(pth2)
else:
    print('Somthing is rong')

In [3]:
static_borrower.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Выведем информацию по датафрейму **static_borrower**, а так же проверим правильность написания наименований колонок.

In [4]:
static_borrower.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В наименованиях колонок ошибок нет, из ТЗ заказчика нам известно следующее описание столбцов:  
`children` — количество детей в семье    
`days_employed` — общий трудовой стаж в днях    
`dob_years` — возраст клиента в годах    
`education` — уровень образования клиента    
`education_id` — идентификатор уровня образования    
`family_status` — семейное положение    
`family_status_id` — идентификатор семейного положения    
`gender` — пол клиента    
`income_type` — тип занятости    
`debt` — имел ли задолженность по возврату кредитов    
`total_income` — ежемесячный доход   
`purpose` — цель получения кредита    

## Шаг 2.1 Заполнение пропусков <a name="Шаг 2.1 Заполнение пропусков"></a> 

Выполним предобработку данных, для начала проверим наличие пропущенных значений в датафрейме **static_borrower**. Из информации по датафрейму видно, что нулевые значения имеются в столбцах **days_employed**, **total_income**, посчитаем их с помощью функции **_isna()_**

In [5]:
static_borrower.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Определим долю пропущенных значений в каждом из столбцов с пропусками, разделив количество пропусков **static_borrower['days_employed'].isna().sum()** на общее количество значений **static_borrower.shape[0]**:

In [6]:
print('Доля пропущенных значений в столбце days_employed = {:.2%}'
      .format(static_borrower['days_employed'].isna().sum()/static_borrower.shape[0]))

print('Доля пропущенных значений в столбце total_income = {:.2%}'
      .format(static_borrower['total_income'].isna().sum()/static_borrower.shape[0]))

Доля пропущенных значений в столбце days_employed = 10.10%
Доля пропущенных значений в столбце total_income = 10.10%


Доля пропущенных значений в столбцах составляет 10.1 %, что является достаточно весомым, удалить их мы не можем поэтому заполним.    
Данные пропуски обусловлены скорее всего техническими факторами, на что указывает одинаковое количество пропусков в двух столбцах **days_employed** и **total_income**.    
Данные колонки содержат количественные значения, заполним пропущенные ячейки медианным значением посчитанным по оставшимся заполненным значениям. Но до этого необходимо исправить отрицательные значения в столбце **days_employed**, перезаписав его значениями по модулю. Так как отрицательного рабочего стажа не бывает, его либо нет и он равен 0, либо он имеет значение в кол-ве дней.

In [7]:
static_borrower['days_employed'] = static_borrower['days_employed'].abs() #перезапись столбца days_employed значениями по модулю
static_borrower.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Теперь найдём медианные значения среди заполненных ячеек в столбцах **days_employed** и **total_income** и сохраним их в переменные **days_employed_median** и **total_income_median** соответственно, затем выведем на экран их значения.    
Столбец из заполненных значений получим применив фильтр **static_borrower['days_employed'].isna() == False**, потому что функция isna() возвращает False, если значения в ячейках отличны от NaN или None.

In [8]:
days_employed_median = static_borrower.loc[static_borrower['days_employed']\
                                      .isna() == False,'days_employed'].median()

print('Медианное значение для заполнение пропусков в столбце days_employed =', days_employed_median)

total_income_median = static_borrower.loc[static_borrower['total_income']\
                                     .isna() == False, 'total_income'].median()

print('Медианное значение для заполнение пропусков в столбце total_income =', total_income_median)

Медианное значение для заполнение пропусков в столбце days_employed = 2194.220566878695
Медианное значение для заполнение пропусков в столбце total_income = 145017.93753253992


Заполним пропущенные значения в столбцах **days_employed** и **total_income** значениями **days_employed_median** и **total_income_median** соответственно и перезапишем датафрейм **static_borrower**

In [9]:
#заполнение колонки days_employed
static_borrower['days_employed'] = static_borrower['days_employed'].fillna(days_employed_median) 
#заполнение колонки days_employed
static_borrower['total_income'] = static_borrower['total_income'].fillna(total_income_median) 

Выведем информацию по исправленному датафрейму и проверим наличие нулевых значений в обновлённых колонках.

In [10]:
static_borrower.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Шаг 2.2 Проверка данных на аномалии и исправления<a name="4.2  Шаг 2.2 Проверка данных на аномалии и исправления"></a>

Проверим наш датафрейм на аномалии, тоесть на те значения:    
- которых там быть не должно,но они подходят по типу данных    
- которые значительно отличаются от значений в столбце и могут повлиять на статистику    
- которые не подходят по логике других столбцов

Для этого выведем имеющиеся значения и их количество в солбцах с категориальными значениями методом **value_counts()**. А для столбцов количественных значений применим сортировку в порядке возрастания/убывания методом **sort_values**. 

### Проверка и исправление столбца children

In [11]:
#столбец children
static_borrower['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В столбце **children** имеются аномальные данные, а именно:    
- число детей 20, что маловероятно, скорее всего здесь имеет место быть опечатка, испарвим это значение на 2    
- число детей -1, тоже опечатка, исправим значение на 1    

Заменим данные значения методом **replace()** и снова проверим все уникальные значения в столбце **children**

In [12]:
static_borrower['children'] = static_borrower['children'].replace(20,2) # замена 20 на 2
static_borrower['children'] = static_borrower['children'].replace(-1,1) # замена -1 на 1
#вывод всех уникальных значений в столбце
static_borrower['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Аномалии в столбце **children** исправлены, следующий столбец **dob_years**.

### Проверка и исправление столбцов dob_years и income_type

In [13]:
#столбец dob_years
static_borrower['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

Как видно из вывода функции, в датафрейме имеются люди с возрастом 0 лет, что не возможно, это очередная аномалия. Эти значения можно заполнить медианным значением, но для того, что бы значения соответствовали действительности и логически подходили к другим столбцам, это не подходит, потому что в датафрейме имеются пенсионеры. Поэтому нули для занятых сотрудников заполним медианным значением из выборки по занятым сотрудникам, а для пенсионеров нули зполним медианным значением среди выборки пенсионеров. Но перед тем как искать медиану для той или иной категории граждан, необходимо в столбце **income_type** проверить регистры и уникальные значения. 

In [14]:
static_borrower['income_type'].value_counts() # поиск уникальных значений столбца income_type 

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Поиск уникальных значений по столбцу **income_type** показал, что регистр в столбце одинаковый, ошибок форматирования нет, а значит фильтры по категориям заёмщиков для поиска медианных значений охватит все необходимые категории.

In [15]:
for_retiree_median = static_borrower.loc[static_borrower['income_type'] == 'пенсионер', 'dob_years']\
                                    .median()
print('Медианное значение для заполнение нулей в столбце dob_years для пенсионеров =', \
      for_retiree_median)
busy_median = static_borrower.loc[static_borrower['income_type'] != 'пенсионер', 'dob_years']\
                             .median()
print('Медианное значение для заполнение нулей в столбце dob_years для занятых =', busy_median)

Медианное значение для заполнение нулей в столбце dob_years для пенсионеров = 60.0
Медианное значение для заполнение нулей в столбце dob_years для занятых = 39.0


Заменим нулевые значения медианными и проверим их наличие в датафрейме

In [16]:
# замена нулей для пенсионеров
static_borrower.loc[(static_borrower['dob_years'] == 0) & \
                    (static_borrower['income_type'] == 'пенсионер') , 'dob_years'] = \
                     static_borrower['dob_years'].replace(0,for_retiree_median)
# замена нулей для занятых
static_borrower.loc[(static_borrower['dob_years'] == 0) & (static_borrower['income_type'] != 'пенсионер') , 'dob_years'] = \
static_borrower['dob_years'].replace(0,busy_median)
# поиск нулей в столбце dob_years
print('Минимальный возраст заёмщика = ' ,static_borrower['dob_years'].min())

Минимальный возраст заёмщика =  19


### Проверка и исправление столбцов education и education_id

In [17]:
static_borrower[['education', 'education_id']].value_counts()

education            education_id
среднее              1               13750
высшее               0                4718
СРЕДНЕЕ              1                 772
Среднее              1                 711
неоконченное высшее  2                 668
ВЫСШЕЕ               0                 274
Высшее               0                 268
начальное            3                 250
Неоконченное высшее  2                  47
НЕОКОНЧЕННОЕ ВЫСШЕЕ  2                  29
НАЧАЛЬНОЕ            3                  17
Начальное            3                  15
ученая степень       4                   4
УЧЕНАЯ СТЕПЕНЬ       4                   1
Ученая степень       4                   1
dtype: int64

Приведём столбец **education** к нижнему регистру, затем стова проверим уникальные значения

In [18]:
static_borrower['education'] = static_borrower['education'].str.lower()
static_borrower['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Проверим следующие столбцы **family_status** и **family_status_id**

### Проверка и исправление столбцов family_status и family_status_id

In [19]:
static_borrower[['family_status', 'family_status_id']].value_counts()

family_status          family_status_id
женат / замужем        0                   12380
гражданский брак       1                    4177
Не женат / не замужем  4                    2813
в разводе              3                    1195
вдовец / вдова         2                     960
dtype: int64

Здесь всё в порядке, повторений нет, логика значений столбца **family_status** не противоречит толбцу **family_status_id**.    
Перейдём к следующему столбцу **gender**

### Проверка и исправление столбца gender

In [20]:
static_borrower['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

В результате имеем третий вид пола XNA и у нас есть два варианта, либо удалить его, учитывая что он 1 на 21500 значений, либо оставить как есть. Я выбрал первый варинт.

In [21]:
static_borrower = static_borrower.loc[static_borrower['gender'] != 'XNA']
static_borrower['gender'].value_counts()

F    14236
M     7288
Name: gender, dtype: int64

Переходим к следующему столбцу **purpose** 

### Проверка и исправление столбцa purpose

In [22]:
static_borrower['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      623
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

В данном столбце аномалий нет, здесь имеется одинаковые по смыслу, но разные по написанию значения, тоесть неявные дубликаты, удаление не явных дубликот займёмся в п.5 Шаге 2.4

Теперь перейдём к столбцам с количественными значениями **days_employed** и **total_income**. С помощью функции describe() выведем информацию о их значениях.

### Проверка и исправление столбцов days_employed и total_income

In [23]:
static_borrower[['days_employed','total_income']].describe()

days_employed  total_income
count   21524.000000  2.152400e+04
mean    60380.728303  1.651577e+05
std    133260.067321  9.786799e+04
min        24.141633  2.066726e+04
25%      1025.593536  1.077960e+05
50%      2194.220567  1.450179e+05
75%      4779.672511  1.955387e+05
max    401755.400475  2.265604e+06

К ежемесечным доходу вопросов нет, чего не скажешь о трудовом стаже. Трудовое законодательство РФ позволяет работать с 16 лет. Определим максимальный возраст из датафрейма и проверим максимально возможное количество дней трудового стажа. Затем выясним сколько позиций выходят за рамки этого значения.

In [24]:
max_old = static_borrower['dob_years'].max() #самый старый в датафрейме
max_days_employed = (max_old - 16) * 365  # максимально возможное количество дней трудового стажа
display(static_borrower.loc[static_borrower['days_employed'] > max_days_employed]) # все позиции в которых трудовой стаж 
#превышает максимально возможный

children  days_employed  dob_years education  education_id  \
4             0  340266.072047         53   среднее             1   
18            0  400281.136913         53   среднее             1   
24            1  338551.952911         57   среднее             1   
25            0  363548.489348         67   среднее             1   
30            1  335581.668515         62   среднее             1   
...         ...            ...        ...       ...           ...   
21505         0  338904.866406         53   среднее             1   
21508         0  386497.714078         62   среднее             1   
21509         0  362161.054124         59    высшее             0   
21518         0  373995.710838         59   среднее             1   
21521         0  343937.404131         67   среднее             1   

               family_status  family_status_id gender income_type  debt  \
4           гражданский брак                 1      F   пенсионер     0   
18            вдовец / вдова                 2      F   пенсионер     0   
24     Не женат / не замужем                 4      F   пенсионер     0   
25           женат / замужем                 0      M   пенсионер     0   
30           женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
21505       гражданский брак                 1      M   пенсионер     0   
21508        женат / замужем                 0      M   пенсионер     0   
21509        женат / замужем                 0      M   пенсионер     0   
21518        женат / замужем                 0      F   пенсионер     0   
21521        женат / замужем                 0      F   пенсионер     0   

        total_income                                purpose  
4      158616.077870                        сыграть свадьбу  
18      56823.777243     на покупку подержанного автомобиля  
24     290547.235997  операции с коммерческой недвижимостью  
25      55112.757732                   покупка недвижимости  
30     171456.067993  операции с коммерческой недвижимостью  
...              ...                                    ...  
21505   75439.993167                        сыграть свадьбу  
21508   72638.590915                           недвижимость  
21509   73029.059379               операции с недвижимостью  
21518  153864.650328                   сделка с автомобилем  
21521  155999.806512                   сделка с автомобилем  

[3445 rows x 12 columns]

3445 позиций имеют завышенный трудовой стаж, которого не может быть. Приравняем все эти значения к максимально возможному.

In [25]:
static_borrower.loc[static_borrower['days_employed'] > max_days_employed, 'days_employed'] = max_days_employed

## Шаг 2.3. Изменение типов данных.

В столбцах **days_employed**, **dob_years**  и **total_income** видим(п.1 Шаг 1) значения типа float. Данные значения переведём в целочисленные типа int.

In [26]:
static_borrower[['days_employed','dob_years','total_income']] = \
static_borrower[['days_employed','dob_years','total_income']].astype('int')

## Шаг 2.4. Удаление дубликатов.

Для начала определим наличие и количество явных дубликатов с помощью функции **duplicated().sum()**

In [27]:
print('Количество явных дубликатов =', static_borrower.duplicated().sum())

Количество явных дубликатов = 71


Удалим явные дубликаты из датафрейма и изменим индексацию

In [28]:
static_borrower = static_borrower.drop_duplicates().reset_index(drop=True)
print('Количество явных дубликатов =', static_borrower.duplicated().sum())

Количество явных дубликатов = 0


С явными дубликатами мы разобрались, остались не явные. По результатам шага 2.2 нам известно, что дубликаты остались только в столбце **purpose**, где имеются одинаковые по смыслу но разные по написанию цели кредита. 

In [29]:
# функция категоризации целей кредита
def category_purpose(purpose):
    if  'авто' in purpose:
        return 'операции с автомобилем'
    elif 'недвиж' in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образо' in purpose:
        return 'получение образования'

Перезапишем столбец **purpose** c применением функции **category_purpose**

In [30]:
static_borrower['purpose'] = static_borrower['purpose'].apply(category_purpose)
static_borrower['purpose'].value_counts()

операции с недвижимостью    10810
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2324
Name: purpose, dtype: int64

## Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Перед декомпозицией исходного датафрейма, сформирем дополнительные датафреймы из столбцов **education_id** и **education** и столбцов **family_status_id** и **family_status**, для сохранения инфорации в них.

In [31]:
save_education = static_borrower[['education', 'education_id']] #новый датафрейм словарь из столбцов education_id и education
save_family_status = static_borrower[['family_status', 'family_status_id']] #новый датафрейм словарь из столбцов education_id и education

Теперь можем удалить столбцы **education** и **family_status** из исходного датафрейма

In [32]:
static_borrower = static_borrower.drop(['education', 'family_status'], axis = 1)

## Шаг 2.6. Категоризация дохода.

Распределим доходы потенциальных заёмщиков по категориям:    
- 0–30000 — **E**;
- 30001–50000 — **D**;    
- 50001–200000 — **C**;    
- 200001–1000000 — **B**;    
- 1000001 и выше — **A**   

Для этого напишем функцию **category_income**, которая будет определять категорию дохода в зависимости от значения в столбце **total_income**, проверим её работоспособность.

In [33]:
def category_income(income):
    if income <= 30000:
        return 'E'
    if 30000 < income <= 50000:
        return 'D'
    if 50000 < income <= 200000:
        return 'C'
    if 200000 < income <= 1000000:
        return 'B'
    if income > 1000000:
        return 'A'
print('Заработная плата 250000 это категория', category_income(250000))

Заработная плата 250000 это категория B


Создадим новый столбец **total_income_id** после столбца **total_income** в нашем датафрейме

In [34]:
static_borrower.insert(9, 'total_income_id', static_borrower['total_income'].apply(category_income))

Сохраним стобцы **total_income_id** и **total_income** в новый датафрейм **save_total_income**, а затем удалим столбец **total_income**

In [35]:
#новый датафрейм словарь из столбцов total_income и total_income_id
save_total_income = static_borrower[['total_income', 'total_income_id']] 
#удаление столбца total_income из исходного датафрейма
static_borrower = static_borrower.drop('total_income', axis=1)
static_borrower.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          21535         53             1                 1      F   

  income_type  debt total_income_id                   purpose  
0   сотрудник     0               B  операции с недвижимостью  
1   сотрудник     0               C    операции с автомобилем  
2   сотрудник     0               C  операции с недвижимостью  
3   сотрудник     0               B     получение образования  
4   пенсионер     0               C        проведение свадьбы

## Шаг 2.7. Категоризация целей кредита.

После удаления неявных дубликатов из столбца **purpose**(п.5 Шаг 2.4), в нём остались только категоризированные значения, дальше разделять на категории не имеет смысла.

## Ответы на вопросы.

### Вопрос 1:  Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для ответа на этот вопрос сгруппируем с помощью метода **pivot_table** датафрейм **static_borrower** по столбцу **children** значения для подсчёта суммы и кол-ва возьмём из столбца **debt**

In [36]:
frame_of_children = static_borrower.pivot_table(index = ['children'], values = 'debt',\
                                                aggfunc = ['sum','count'])
frame_of_children

sum  count
          debt   debt
children             
0         1063  14090
1          445   4855
2          202   2128
3           27    330
4            4     41
5            0      9

У нас получился датафрейм **frame_of_children**, добавим в него столбец **share** с расчётом доли должников в зависимости от наличия и количества детей

In [37]:
frame_of_children['share'] = (frame_of_children['sum']/frame_of_children['count'])*100
frame_of_children = frame_of_children.round(1)
# объединим наименования столбцов таблицы
frame_of_children.columns = ['sum_debt', 'count_debt', 'share']
# отсортируем по столбцу share
frame_of_children.sort_values(by='share')

sum_debt  count_debt  share
children                             
5                0           9    0.0
0             1063       14090    7.5
3               27         330    8.2
1              445        4855    9.2
2              202        2128    9.5
4                4          41    9.8

### Вывод 1

Как видно из последнего столбца **share** датафрейма **frame_of_family** доля должников у заёмщиков без детей меньше и составляет 7.5 %. Тоесть у тех людей, у которых нет детей реже возникают задолжности по платежам, что объяснимо, исходя из того, что трат у таких людей меньше. Долю должников для родителей 5х детей не рассматривает, так как для статистики таких родителей мало, всего 9 из 21454.

### Вопрос 2:  Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для ответа на этот вопрос сгруппируем с помощью метода **pivot_table** датафрейм **static_borrower** по столбцу **family_status_id** значения для подсчёта суммы и кол-ва возьмём из столбца **debt**

In [38]:
frame_of_family = static_borrower.pivot_table(index = ['family_status_id'], values = 'debt', aggfunc = ['sum','count'])
frame_of_family

sum  count
                 debt   debt
family_status_id            
0                 931  12339
1                 388   4150
2                  63    959
3                  85   1195
4                 274   2810

У нас получился датафрейм **frame_of_family**, добавим в него столбец **share** с расчётом доли должников в зависимости от семейного положения.

In [39]:
frame_of_family['share'] = (frame_of_family['sum']/frame_of_family['count'])*100
frame_of_family = frame_of_family.round(1)
# объединим наименования столбцов таблицы
frame_of_family.columns = ['sum_debt', 'count_debt', 'share']
# отсортируем по столбцу share
frame_of_family = frame_of_family.sort_values(by='share')

Объединим датафрейм **frame_of_family** с таблицей словарём **save_family_status**, при этом удалим дубликаты из датафрейма **save_family_status** без его перезаписи **save_family_status.drop_duplicates().reset_index(drop=True)**

In [40]:
frame_of_family = frame_of_family.merge(save_family_status.drop_duplicates().reset_index(drop=True), on='family_status_id', how = 'left')
frame_of_family[['family_status', 'share']]

family_status  share
0         вдовец / вдова    6.6
1              в разводе    7.1
2        женат / замужем    7.5
3       гражданский брак    9.3
4  Не женат / не замужем    9.8

### Вывод 2

Как видно из результатов исследования, семейное положение влияет на долю должников. Так наименьший процент должников среди вдов и вдовцов - 6.6 %, возможно потому, что уже имеется нажитое вместе имущество, расходы не велики на развлечения и прочие радости жизни, в основном это наверняка пожилые, зрелые люди, отлично умеющие распоряжаться своими доходами.    
Наибольший процент должников среди не женатых/ не замужних 9.8 %, скоре всего здесь большая часть молодых, ещё не очень умеющих распоряжаться своими доходами людей у которых траты на повседневную жизнь больше нежели у остальных.

### Вопрос 3:  Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для ответа на этот вопрос сгруппируем с помощью метода **pivot_table** датафрейм **static_borrower** по столбцу **total_income_id** значения для подсчёта суммы и кол-ва возьмём из столбца **debt**

In [41]:
frame_of_income = static_borrower.pivot_table(index = ['total_income_id'], values = 'debt', aggfunc = ['sum','count'])
frame_of_income

sum  count
                 debt   debt
total_income_id             
A                   2     25
B                 356   5040
C                1360  16016
D                  21    350
E                   2     22

У нас получился датафрейм **frame_of_income**, добавим в него столбец **share** с расчётом доли должников в зависимости от семейного положения

In [42]:
frame_of_income['share'] = (frame_of_income['sum']/frame_of_income['count'])*100
frame_of_income = frame_of_income.round(1)
# объединим наименования столбцов таблицы
frame_of_income.columns = ['sum_debt', 'count_debt', 'share']
# отсортируем по столбцу share
frame_of_income = frame_of_income.sort_values(by='share')
frame_of_income

sum_debt  count_debt  share
total_income_id                             
D                      21         350    6.0
B                     356        5040    7.1
A                       2          25    8.0
C                    1360       16016    8.5
E                       2          22    9.1

### Вывод 3

Как видно из исследования наименьшая доля должников 6 % у людей с уровнем дохода D (30001–50000 у.е. в месяц). Наибольшее количество должников 9.1 % с уровнем дохода E (до 30000 у.е. в месяц). Это объяснимо, потому что это самый низкий доход из нашего датафрейма. Но выводы по этому вопросу не могут быть однозначны, так как данных по уровням дохода А,Е мало(25,22 человек это около 0.1 % от общего количества).

### Вопрос 4:  Как разные цели кредита влияют на его возврат в срок?

Для ответа на этот вопрос сгруппируем с помощью метода **pivot_table** датафрейм **static_borrower** по столбцу **purpose** значения для подсчёта суммы и кол-ва возьмём из столбца **debt**

In [43]:
frame_of_purpose = static_borrower.pivot_table(index = ['purpose'], values = 'debt', aggfunc = ['sum','count'])
frame_of_purpose

sum  count
                         debt   debt
purpose                             
операции с автомобилем    403   4306
операции с недвижимостью  782  10810
получение образования     370   4013
проведение свадьбы        186   2324

У нас получился датафрейм **frame_of_purpose**, добавим в него столбец **share** с расчётом доли должников в зависимости от семейного положения

In [44]:
frame_of_purpose['share'] = (frame_of_purpose['sum']/frame_of_purpose['count'])*100
frame_of_purpose = frame_of_purpose.round(1)
# объединим наименования столбцов таблицы
frame_of_purpose.columns = ['sum_debt', 'count_debt', 'share']
# отсортируем по столбцу share
frame_of_purpose = frame_of_purpose.sort_values(by='share')
# выведем интересующий нас результат
frame_of_purpose['share']

purpose
операции с недвижимостью    7.2
проведение свадьбы          8.0
получение образования       9.2
операции с автомобилем      9.4
Name: share, dtype: float64

### Вывод 4

Как видно из результата исследования, лучше всех отдают кредит люди берущие деньги на недвижимость и строительство, доля должников у обоих категорий 7.2 %. Что объяснимо, так как это жизненно необходимые потребности в жилье, к этому люди, в большинстве своём, идут обдуманно, взвесив свои доходы и просрочка по оплате может дорого стоить. Наибольший процент должников у тех, кто берёт кредит на автомобиль, возможно потому, что сумма на автомобиль достаточно велика в отличии от свадьбы и образования, а просрочка по платежам не несёт больших рисков остаться без жизненно необходимого.

## Общий вывод:

Основываясь на полученные данные можно прийти к следующему выводу: Семейное положение и наличие детей в семье безусловно влияют на факт погашения кредита. Так мы можем увидеть, что вероятность клиента имеющего 4 ребёнка стать должником составляет 9.8%, а вероятность клиента принадлежащего к группе "не женат / не замужем" стать должником составляет 9.8%. Поэтому логично будет предположить, что наиболее безопасными для банка заемщиками являются семьи без детей и люди являющиеся вдовой/вдовцом.